src/server/pps/server/s3g_sidecar.go

```diff
@@ -0,0 +1,475 @@
+				return // Nothing to delete; this isn't an s3 pipeline (shouldn't happen)
+			}
+
+			if err := backoff.RetryNotify(func() error {
+				err := s.apiServer.env.GetKubeClient().CoreV1().Services(s.apiServer.namespace).Delete(
+					ppsutil.SidecarS3GatewayService(jobID),
+					&metav1.DeleteOptions{OrphanDependents: new(bool) /* false */})
+				if err != nil && strings.Contains(err.Error(), "not found") {
+					return nil // service already deleted
+				}
+				return err
+			}, backoff.NewExponentialBackOff(), func(err error, d time.Duration) error {
+				logrus.Errorf("error deleting kubernetes service for s3 %q gateway sidecar: %v; retrying in %v", jobID, err, d)
+				return nil
+			}); err != nil {
+				logrus.Errorf("permanent error deleting kubernetes service for %q s3 gateway sidecar: %v", jobID, err)
+			}
+		},
+	}
+}
+
+func (h *jobHandler) start() {
+	defer func() {
+		logrus.Errorf("sidecar s3 gateway: start() is exiting; this should not happen")
+		panic(
+			fmt.Sprintf("!sidecar s3 gateway start() is exiting; this should never happen"),
+		)
+	}()
+establish_watch:
+	for { // reestablish watch in a loop, in case there's a watch error
+		var watcher watch.Watcher
+		backoff.Retry(func() error {
+			var err error
+			watcher, err = h.s3g.apiServer.jobs.ReadOnly(context.Background()).WatchByIndex(
+				ppsdb.JobsPipelineIndex, h.s3g.pipelineInfo.Pipeline)
+			if err != nil {
+				return fmt.Errorf("error creating watch: %v", err)
+			}
+			return nil
+		}, backoff.NewInfiniteBackOff())
+		defer watcher.Close()
+
+		for e := range watcher.Watch() {
+			jobID := string(e.Key)
+			if e.Type == watch.EventError {
+				logrus.Errorf("sidecar s3 gateway watch error: %v", e.Err)
+				continue establish_watch
+			}
+
+			// create new ctx for this job
+			jobCtx, jobCancel := context.WithCancel(context.Background())
+			h.processJobEvent(jobCtx, e.Type, jobID)
+			// spin off handler for job termination. 'watcher' will not see any job
+			// state updates after the first because job state updates don't update
+			// the pipelines index, so this establishes a watcher that will.
+			go h.end(jobCtx, jobCancel, jobID)
+		}
+	}
+}
+
+// end watches 'jobID' and calls h.OnTerminate() when the job finishes.
+func (h *jobHandler) end(ctx context.Context, cancel func(), jobID string) {
+	defer cancel()
+establish_watch:
+	for { // reestablish watch in a loop, in case there's a watch error
+		var watcher watch.Watcher
+		backoff.Retry(func() error {
+			var err error
+			watcher, err = h.s3g.apiServer.jobs.ReadOnly(ctx).WatchOne(jobID)
+			if err != nil {
+				return fmt.Errorf("error creating watch: %v", err)
+			}
+			return nil
+		}, backoff.NewInfiniteBackOff())
+		defer watcher.Close()
+
+		for e := range watcher.Watch() {
+			jobID := string(e.Key)
+			if e.Type == watch.EventError {
+				logrus.Errorf("sidecar s3 gateway watch job %q error: %v", jobID, e.Err)
+				continue establish_watch
```

Comment:
Same as above: I found this easier to follow visually when I was writing it because of the `if` inside the `for` but I don't feel strongly so I changed it.